tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorContainer.java

```diff
@@ -0,0 +1,52 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+/**
+ * An interface representing a collection or group of tensors.  Provides methods for resource management.
+ */
+public interface TensorContainer extends AutoCloseable {
+
+  /**
+   * Get the tensors held by this object.
+   */
+  Iterable<? extends Tensor> tensors();
+
+  /**
+   * Detach these tensors from any scopes managing them.  They must be manually closed or attached to another scope.
+   *
+   * @see Tensor#detach()
+   */
+  default void detach() {
```

Comment:
The `ResourceScope` currently being proposed by Panama doesn't have a way to "detach" objects. They are reusing the linked list that is also being used by the GC to track references, and I think that for efficiency, it doesn't have a way to remove objects. In the case of something like JavaCPP where we may need to map C++ concepts like `std::shared_ptr`, that's problematic, but in this case, we can safely do without. One can simply create a new "view" of a `Tensor`, and that's it. The original `Tensor` can be safely discarded without having to detach or do anything else. Removing `detach()` also makes `attach()` redundant, further simplifying the design. To make this work though we need something like reference counting at the lower level. Fortunately, we already have that with JavaCPP. :)

Reply:
By "new view of a tensor" do you mean something like a new `RawTensor` but with the same native handle?  Where instead of moving tensors between scopes w/ attach/detach you would just create a new view in the new scope?

All that would do is change `parent.attach(tensor)` to `tensor.inScope(parent)` which I don't see a point to (it would also cause issues w/ collections).  You'd need to track scope membership anyways for closing, so there wouldn't be any performance improvement.  Removals and insertions are constant time anyways, it's a set.

We could get rid of detach at an API level anyways if we want to prevent free/unscoped tensors, but there could be edge cases where it's useful and it's not dangerous imo to warrant removal.

Reply:
> By "new view of a tensor" do you mean something like a new `RawTensor` but with the same native handle? Where instead of moving tensors between scopes w/ attach/detach you would just create a new view in the new scope?

Yes, but we can start by just making copies of the data, and only worry about that when we encounter a situation where is becomes unacceptable (either for performance reasons or for other reasons). TF Core itself already copies data all over the place all the time. It's not exactly fast to begin with. I really doubt we will have to worry about this. For starters, let's just have users copy their data around and see where that goes?

> All that would do is change `parent.attach(tensor)` to `tensor.inScope(parent)` which I don't see a point to (it would also cause issues w/ collections). You'd need to track scope membership anyways for closing, so there wouldn't be any performance improvement. Removals and insertions are constant time anyways, it's a set.

I meant `attach()` as a public API. Users shouldn't need to worry about calling it manually. If we decide it's something that needs to be called only inside `Tensor` during construction, we could have it, for example, as a package private method, or by having a `Scope` interface with essentially only the `close()` method and making the implementation of `TensorScope` itself private, or maybe some other way. IMO, that simplifies the API a lot.

> We could get rid of detach at an API level anyways if we want to prevent free/unscoped tensors, but there could be edge cases where it's useful and it's not dangerous imo to warrant removal.

That's not the point though. The point is to not add features we may not need and that could cause performance degradation in the future when trying to use new stuff in the JDK that may not allow us to do that. We can always add new stuff later, when we have a better understanding of what we do need.

Reply:
Since we're only publishing "alpha" releases at this point though, we can also leave it that way for now, and see if users actually use `attach()` and/or `detach()` and how they feel about it. We can remove stuff before "beta" releases. I'd actually be fine to merge it like that.

Reply:
That seems fine.  The main use case I was looking to support is subscopes, i.e. confining an expensive operation:

Lets say we have some method that creates a lot of short lived temporary tensors:
```java
public static TFloat32 withLotsOfTemps(TensorScope scope){ ... }
```
The using it would look like
```java
public static TFloat32 buildTensor(TensorScope scope, int size){
  TFloat32 sum = withLotsOfTemps(scope);
  // ...
  return sum;
}
```
This is non-optimal because the large amount of temporary tensors live as long as the outer scope, there should be a way to further confine their lifetime, i.e.:
```java
public static TFloat32 buildTensor(TensorScope scope, int size){
  try(TensorScope innerScope = new TensorScope()){
    TFloat32 sum = withLotsOfTemps(scope);
    // ...
    return sum;
  }
}
```
However we need a way to extract the result from the inner scope.  Since we have the outer scope, we would want to return `scope.attach(sum)`.  And if we support not specifying a scope I could see wanting to manually add the result to a scope later (i.e. if the method you're calling uses the no-scope version for whatever reason but you have a scope you want to use).

Similarly, for detach, if we use the `EagerSession` to scope eager `asTensor()` tensors, I think it would be fairly common to want a tensor (or NDArray created from it) to live longer than the session that calculated it.  If we do use optional scopes I could see getting rid of detach, but I prefer required scopes + detach as it doesn't require so many duplicated methods and doesn't require APIs to support it.

This can all be done with copying, of course, but that's expensive and it doesn't seem likely that whatever performance costs this could cause by differing from Panama would be worse (worse case scenario is just copying it into a new `ResourceScope` afaik).  It would be good to see how common this actually is.